src/pyproject_fmt/formatter/pep508.py

```diff
@@ -1,63 +1,48 @@
 from __future__ import annotations
 
-import re
-
+from packaging.requirements import Requirement
+from tomlkit.api import string as toml_string
 from tomlkit.items import Array, String
 
-from .util import ArrayEntries, sorted_array
-
-BASE_NAME_REGEX = re.compile(r"[^!=><~\s@]+")
-REQ_REGEX = re.compile(r"(===|==|!=|~=|>=?|<=?|@)\s*([^,]+)")
-
-
-def _req_base(lib: str) -> str:
-    match = re.match(BASE_NAME_REGEX, lib)
-    if match is None:
-        raise ValueError(repr(lib))
-    return match.group(0)
-
-
-def _normalize_lib(lib: str) -> str:
-    base = _req_base(lib)
-    values = sorted(
-        (f"{m.group(1)}{m.group(2)}" for m in REQ_REGEX.finditer(lib)), key=lambda c: ("<" in c, ">" in "c", c)
-    )
-    if values:  # strip .0 version
-        while values[0].endswith(".0") and (values[0].startswith(">=") or values[0].startswith("==")):
-            values[0] = values[0][:-2]
-    return f"{base}{','.join(values)}"
+from .util import sorted_array
 
 
 def normalize_req(req: str) -> str:
-    lib, _, envs = req.partition(";")
-    normalized = _normalize_lib(lib)
-    envs = envs.strip()
-    if not envs:
-        return normalized
-    return f"{normalized};{envs}"
+    parsed = Requirement(req)
+    for spec in parsed.specifier:
+        if spec.operator in (">=", "=="):
+            version = spec.version
+            while version.endswith(".0"):
+                version = version[:-2]
+            spec._spec = (spec._spec[0], version)
+    return str(parsed)
 
 
 def normalize_requires(raws: list[str]) -> list[str]:
     values = (normalize_req(req) for req in raws if req)
-    normalized = sorted(values, key=lambda req: (";" in req, _req_base(req), req))
+    normalized = sorted(values, key=lambda req: (";" in req, Requirement(req).name, req))
     return normalized
 
 
-def _get_pkg_name(entry: ArrayEntries) -> str:
-    match = BASE_NAME_REGEX.match(entry.text)
-    assert match is not None
-    return match.group(0)
+def _best_effort_string_repr(req: str) -> String:
+    # Convert requirement to a TOML string, choosing the most appropriate representation (basic or literal).
+    # This function will attempt to use literal strings to avoid escaping double-quotes ("), if the requirement value
+    # allows it, e.g. it does not contain other reserved characters such as single quotes (') or newlines (\\n).
+    try:
+        return toml_string(req, literal=('"' in req))
+    except ValueError:
+        return toml_string(req)
```

Comment:
@abravalheri can you add a test that covers this line?

Reply:
Now that all the quotes are being normalised by `packaging` I think this is check became unnecessary.

I tried to create a fabricated example in https://github.com/tox-dev/pyproject-fmt/blob/ba03716c00d8d18a7b0edf339e08cb17921f0297/tests/formatter/test_project.py#L40

(Mixed quotes due to their presence in the URL is the only situation I can think of when this situation can happen.

I checked with curl and the browsers and this kind of URL seems to be accepted, e.g `curl "https://en.wikipedia.org/w/index.php?title=''"`)

I don't know if you want to keep this or just go for the simpler approach that is not catching the exception.

Reply:
We can keep it as is :+1: